mobilecoind-json/src/main.rs

```diff
@@ -0,0 +1,317 @@
+
+/// Command line config, set with defaults that will work with
+/// a standard mobilecoind instance
+#[derive(Clone, Debug, StructOpt)]
+#[structopt(
+    name = "mobilecoind-rest-gateway",
+    about = "A REST frontend for mobilecoind"
+)]
+pub struct Config {
+    /// Host to listen on.
+    #[structopt(long, default_value = "127.0.0.1")]
+    pub listen_host: String,
+
+    /// Port to start webserver on.
+    #[structopt(long, default_value = "9090")]
+    pub listen_port: u16,
+
+    /// MobileCoinD URI.
+    #[structopt(long, default_value = "127.0.0.1:4444")]
+    pub mobilecoind_host: String,
+
+    /// SSL
+    #[structopt(long)]
+    pub use_ssl: bool,
+}
+
+/// Connection to the mobilecoind client
+struct State {
+    pub mobilecoind_api_client: MobilecoindApiClient,
+}
+
+#[derive(Serialize, Default)]
+struct JsonEntropyResponse {
+    entropy: String,
+}
+
+/// Requests a new root entropy from mobilecoind
+#[get("/entropy")]
+fn entropy(state: rocket::State<State>) -> Result<Json<JsonEntropyResponse>, String> {
+    let resp = state
+        .mobilecoind_api_client
+        .generate_entropy(&mc_mobilecoind_api::Empty::new())
+        .map_err(|err| format!("Failed getting entropy: {}", err))?;
+    Ok(Json(JsonEntropyResponse {
+        entropy: hex::encode(resp.entropy),
+    }))
+}
+
+#[derive(Deserialize, Default)]
+struct JsonMonitorRequest {
+    key: String,
+    start: u64,
+    number: u64,
+}
+
+#[derive(Serialize, Default)]
+struct JsonMonitorResponse {
+    monitor_id: String,
+}
+
+/// Creates a monitor. Data for the key and range is POSTed using the struct above.
+#[post("/create-monitor", format = "json", data = "<monitor>")]
+fn create_monitor(
+    state: rocket::State<State>,
+    monitor: Json<JsonMonitorRequest>,
+) -> Result<Json<JsonMonitorResponse>, String> {
+    let entropy =
+        hex::decode(&monitor.key).map_err(|err| format!("Failed to decode hex key: {}", err))?;
+
+    let mut req = mc_mobilecoind_api::GetAccountKeyRequest::new();
+    req.set_entropy(entropy.to_vec());
+
+    let mut resp = state
+        .mobilecoind_api_client
+        .get_account_key(&req)
+        .map_err(|err| format!("Failed getting account key for entropy: {}", err))?;
+
+    let account_key = resp.take_account_key();
+
+    let mut req = mc_mobilecoind_api::AddMonitorRequest::new();
+    req.set_account_key(account_key);
+    req.set_first_subaddress(monitor.start);
+    req.set_num_subaddresses(monitor.number);
+    req.set_first_block(0);
+
+    let monitor_response = state
+        .mobilecoind_api_client
+        .add_monitor(&req)
+        .map_err(|err| format!("Failed adding monitor: {}", err))?;
+
+    Ok(Json(JsonMonitorResponse {
+        monitor_id: hex::encode(monitor_response.monitor_id),
+    }))
+}
+
+#[derive(Serialize, Default)]
+struct JsonBalanceResponse {
+    balance: u64,
+}
+
+/// Balance check using a created monitor and subaddress index
+#[get("/monitors/<monitor_hex>/balance/<index>")]
```

Comment:
subaddress_index?

Reply:
I think flipping from `<type>/<id>` to `<action>/<id2>` is a little strange here, usually this would be something like `/monitors/<hex>/subaddress/<idx>/balance`.

Reply:
Yeah, that makes more sense, I'll flip them